File: scripts/generate_icons.mjs
import fs from "node:fs";
import path from "node:path";
import sharp from "sharp";

const repoRoot = path.resolve(import.meta.dirname, "..");
const publicDir = path.join(repoRoot, "public");
const src = path.join(publicDir, "homeicon.png");

async function main() {
  if (!fs.existsSync(src)) {
    throw new Error(`Source icon not found: ${src}`);
  }

  const base = sharp(src).ensureAlpha();

  const sizes = {
    "favicon-16x16.png": 16,
    "favicon-32x32.png": 32,
    "favicon-48x48.png": 48,
    "apple-touch-icon.png": 180,
    "android-chrome-192x192.png": 192,
    "android-chrome-512x512.png": 512,
  };

  await Promise.all(
    Object.entries(sizes).map(async ([name, size]) => {
      const outPath = path.join(publicDir, name);
      await base
        .clone()
        .resize(size, size, { fit: "cover" })
        .png({ compressionLevel: 9, adaptiveFiltering: true })
        .toFile(outPath);
    }),
  );

  // OG image (1200x630): centered logo on black background
  const ogPath = path.join(publicDir, "og-image.png");
  const logoSize = Math.floor(630 * 0.72);
  const logo = await base
    .clone()
    .resize(logoSize, logoSize, { fit: "contain" })
    .png()
    .toBuffer();

  await sharp({
    create: {
      width: 1200,
      height: 630,
      channels: 4,
      background: { r: 0, g: 0, b: 0, alpha: 1 },
    },
  })
    .composite([
      {
        input: logo,
        left: Math.floor((1200 - logoSize) / 2),
        top: Math.floor((630 - logoSize) / 2),
      },
    ])
    .png({ compressionLevel: 9, adaptiveFiltering: true })
    .toFile(ogPath);

  // Copy to common fallback filenames some browsers request
  fs.copyFileSync(path.join(publicDir, "favicon-32x32.png"), path.join(publicDir, "favicon.png"));

  process.stdout.write(`Wrote icons into ${publicDir}\n`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/generate_icons.py
from __future__ import annotations

from pathlib import Path

from PIL import Image


def main() -> None:
    repo_root = Path(__file__).resolve().parents[1]
    public_dir = repo_root / "public"
    src = public_dir / "homeicon.png"

    img = Image.open(src).convert("RGBA")

    sizes: dict[str, tuple[int, int]] = {
        "favicon-16x16.png": (16, 16),
        "favicon-32x32.png": (32, 32),
        "favicon-48x48.png": (48, 48),
        "android-chrome-192x192.png": (192, 192),
        "android-chrome-512x512.png": (512, 512),
        "apple-touch-icon.png": (180, 180),
    }

    for name, size in sizes.items():
        resized = img.resize(size, Image.Resampling.LANCZOS)
        resized.save(public_dir / name, format="PNG", optimize=True)

    # Multi-size favicon.ico
    img.save(public_dir / "favicon.ico", sizes=[(16, 16), (32, 32), (48, 48)])

    # OG image: centered logo on black background (1200x630)
    og_size = (1200, 630)
    og = Image.new("RGBA", og_size, (0, 0, 0, 255))
    max_w, max_h = og_size
    target_h = int(max_h * 0.72)
    target_w = target_h
    logo = img.resize((target_w, target_h), Image.Resampling.LANCZOS)
    ox = (max_w - target_w) // 2
    oy = (max_h - target_h) // 2
    og.alpha_composite(logo, (ox, oy))
    og.convert("RGB").save(public_dir / "og-image.png", format="PNG", optimize=True)

    print(f"Wrote icons into {public_dir}")


if __name__ == "__main__":
    main()

